File: ZQUIKitChainTool/Objective-C/Category/UIKit+ZQAdd.m
//
//  UIKit+ZQAdd.m
//  ZQUIKitChainTool
//
//  Created by caozhiqiang on 2019/9/29.
//

#import "UIKit+ZQAdd.h"

// 其他视图创建
#define ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(class, method) \
- (class *)method { \
    class *subView = [[class alloc] init]; \
    [self addSubview:subView]; \
    return subView; \
} \

//Gesture
#define ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(class, method) \
- (class *)method { \
    class *subGesture = [[class alloc] init]; \
    [self addGestureRecognizer:subGesture]; \
    return subGesture; \
} \

// BarItem
#define ZQ_UIKIT_CATEGORY_BAR_ADD_IMPLEMENT(class, method) \
+ (class *)method { \
    return [[class alloc] init]; \
} \

@implementation UIView (ZQAdd)

#pragma mark - View

ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIView, addView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UILabel, addLabel);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIImageView, addImageView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIScrollView, addScrollView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UITableView, addTableView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UITextView, addTextView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UITextField, addTextField);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIPickerView, addPickerView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(WKWebView, addWebView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIProgressView, addProgressView);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIActivityIndicatorView, addActivityIndicatorView);

- (UICollectionView *)addCollectionView {
    UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
    UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
    [self addSubview:collectionView];
    return collectionView;
}

#pragma mark - Control

ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIControl, addControl);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIButton, addButton);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIDatePicker, addDatePicker);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UIPageControl, addPageControl);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UISegmentedControl, addSegmentedControl);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UISlider, addSlider);
ZQ_UIKIT_CATEGORY_ADD_IMPLEMENT(UISwitch, addSwitch);

#pragma mark - Gesture

ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UIGestureRecognizer, addGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UITapGestureRecognizer, addTapGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UISwipeGestureRecognizer, addSwipeGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UIScreenEdgePanGestureRecognizer, addScreenGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UIRotationGestureRecognizer, addRotationGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UIPanGestureRecognizer, addPanGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UILongPressGestureRecognizer, addLongPressGesture);
ZQ_UIKIT_CATEGORY_GESTURE_ADD_IMPLEMENT(UIPinchGestureRecognizer, addPinchGesture);

@end


@implementation UIBarItem (ZQAdd)

ZQ_UIKIT_CATEGORY_BAR_ADD_IMPLEMENT(UIBarItem, createBarItem);
ZQ_UIKIT_CATEGORY_BAR_ADD_IMPLEMENT(UIBarButtonItem, createBarButtonItem);
ZQ_UIKIT_CATEGORY_BAR_ADD_IMPLEMENT(UITabBarItem, createTabBarItem);

@end

File: ZQUIKitChainTool/Swift/Category/UIKit+ZQAdd.swift
//
//  UIView+ZQChainHandler.swift
//  ZQUIKitChainToolSwiftDemo
//
//  Created by caozhiqiang on 2019/9/26.
//  Copyright © 2019 caozhiqiang. All rights reserved.
//

import UIKit
import WebKit

// UIView
public extension UIView  {
    
    var addView: UIView {
        let view = UIView(); self.addSubview(view); return view
    }
    
    var addLabel: UILabel {
        let view = UILabel(); self.addSubview(view); return view
    }
    
    var addImageView: UIImageView {
        let view = UIImageView(); self.addSubview(view); return view
    }
    
    var addScrollView: UIScrollView {
        let view = UIScrollView(); self.addSubview(view); return view
    }
    
    var addTableView: UITableView {
        let view = UITableView(); self.addSubview(view); return view
    }
    
    var addTextView: UITextView {
        let view = UITextView(); self.addSubview(view); return view
    }
    
    var addTextField: UITextField {
        let view = UITextField(); self.addSubview(view); return view
    }
    
    var addPickerView: UIPickerView {
        let view = UIPickerView(); self.addSubview(view); return view
    }
    
    var addWebView: WKWebView {
        let view = WKWebView(); self.addSubview(view); return view
    }
    
    var addProgressView: UIProgressView {
        let view = UIProgressView(); self.addSubview(view); return view
    }
    
    var addActivityIndicatorView: UIActivityIndicatorView {
        let view = UIActivityIndicatorView(); self.addSubview(view); return view
    }
    
    var addCollectionView: UICollectionView {
        let layout = UICollectionViewFlowLayout(); let view = UICollectionView(frame: .zero, collectionViewLayout: layout); self.addSubview(view); return view
    }
}



// UIControl
public extension UIView {
    
    var addControl: UIControl {
        let control = UIControl(); self.addSubview(control); return control
    }
    
    var addButton: UIButton {
        let control = UIButton(); self.addSubview(control); return control
    }
    
    var addDatePicker: UIDatePicker {
        let control = UIDatePicker(); self.addSubview(control); return control
    }
    
    var addSegmentedControl: UISegmentedControl {
        let control = UISegmentedControl(); self.addSubview(control); return control
    }
    
    var addSlider: UISlider {
        let control = UISlider(); self.addSubview(control); return control
    }
    
    var addSwitch: UISwitch {
        let control = UISwitch(); self.addSubview(control); return control
    }
    
    var addPageControl: UIPageControl {
        let control = UIPageControl(); self.addSubview(control); return control
    }
}



// Gesture
public extension UIView {
    
    var addGesture: UIGestureRecognizer {
        let gesture = UIGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addTapGesture: UITapGestureRecognizer {
        let gesture = UITapGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addSwipeGesture: UISwipeGestureRecognizer {
        let gesture = UISwipeGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addScreenGesture: UIScreenEdgePanGestureRecognizer {
        let gesture = UIScreenEdgePanGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addRotationGesture: UIRotationGestureRecognizer {
        let gesture = UIRotationGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addPanGesture: UIPanGestureRecognizer {
        let gesture = UIPanGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addLongPressGesture: UILongPressGestureRecognizer {
        let gesture = UILongPressGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
    
    var addPinchGesture: UIPinchGestureRecognizer {
        let gesture = UIPinchGestureRecognizer(); self.addGestureRecognizer(gesture); return gesture
    }
}

public extension UIBarItem {
    
    static var createBarItem: UIBarItem { return UIBarItem() }
    
    static var createBarButtonItem: UIBarButtonItem { return UIBarButtonItem() }
    
    static var createTabBarItem: UITabBarItem { return UITabBarItem() }
}
